File: backend/src/photonflow/blocks/optical/attenuator.py
"""Optical attenuator."""

from __future__ import annotations

from typing import Dict

from photonflow.blocks.base import BaseBlock, register_block
from photonflow.core.signal import Signal
from photonflow.core.sim import SimContext


@register_block("Attenuator")
class Attenuator(BaseBlock):
    PORTS = {"opt_in": "optical", "opt_out": "optical"}
    SPEC = {
        "params": {"loss_db": {"type": "float", "default": 0.0, "unit": "dB"}},
        "nonideal": {"enable": {"type": "bool", "default": False}},
    }

    def process(self, inputs: Dict[str, Signal], ctx: SimContext) -> Dict[str, Signal]:
        opt_in = inputs["opt_in"]
        loss_db = float(self.params.get("loss_db", 0.0))
        data = opt_in.data * (10 ** (-loss_db / 20.0))

        signal = Signal(
            data=data,
            fs=opt_in.fs,
            t0=opt_in.t0,
            center_freq=opt_in.center_freq,
            pol_mode=opt_in.pol_mode,
            meta=dict(opt_in.meta),
        )
        return {"opt_out": signal}

File: photonflow-cpp/src/blocks/optical/attenuator.cpp
/**
 * @file attenuator.cpp
 * @brief Optical attenuator block implementation.
 */

#include "photonflow/blocks/optical/attenuator.hpp"
#include "photonflow/blocks/block_registry.hpp"

#include <cmath>

namespace photonflow {

Attenuator::Attenuator(const std::string &id, const json &params,
                       const json &nonideal)
    : BaseBlock(id, params, nonideal) {}

std::unordered_map<std::string, Signal>
Attenuator::process(const std::unordered_map<std::string, Signal> &inputs,
                    SimContext & /*ctx*/) {
  const Signal &opt_in = inputs.at("opt_in");
  double loss_db = get_param("loss_db", 0.0);

  // Convert dB to linear (field, not power)
  double loss_linear = std::pow(10.0, -loss_db / 20.0);

  Eigen::VectorXcd data = opt_in.data * loss_linear;

  Signal output(data, opt_in.fs, opt_in.t0);
  output.center_freq = opt_in.center_freq;
  output.pol_mode = opt_in.pol_mode;
  output.meta = opt_in.meta;

  return {{"opt_out", std::move(output)}};
}

std::optional<std::string>
Attenuator::port_type(const std::string &port) const {
  if (port == "opt_in" || port == "opt_out")
    return "optical";
  return std::nullopt;
}

json Attenuator::describe() const {
  return {
      {"ports", {{"opt_in", "optical"}, {"opt_out", "optical"}}},
      {"spec",
       {{"params",
         {{"loss_db", {{"type", "float"}, {"default", 0.0}, {"unit", "dB"}}}}},
        {"nonideal", {{"enable", {{"type", "bool"}, {"default", false}}}}}}}};
}

REGISTER_BLOCK(Attenuator, "Attenuator");

} // namespace photonflow
